internal/controllers/clusterextension_controller.go

```diff
@@ -278,6 +286,36 @@ func (r *ClusterExtensionReconciler) installedBundle(ctx context.Context, allBun
 	return resultSet[0], nil
 }
 
+func (r *ClusterExtensionReconciler) validateBundle(bundle *catalogmetadata.Bundle) error {
+	mediaType, err := bundle.MediaType()
+	if err != nil {
+		return fmt.Errorf("unable to determine type of bundle %q: %s", bundle.Name, err)
+	}
+
+	if mediaType != catalogmetadata.MediaTypeRegistry {
+		// We only validate registry+v1 bundles
+		return nil
+	}
+
+	for i := range bundle.Properties {
+		for _, propType := range []string{
+			property.TypePackageRequired,
+			property.TypeGVKRequired,
+			property.TypeConstraint,
+		} {
```

Comment:
Could we instead make this a map and see if the bundle property type exists in the map we return the error?

This might be a premature and low-value optimization, but it reduces the time complexity from O(n*m) to O(n) (where n is the number of bundle properties and m is the unsupported properties) and is, IMO, easier to read

Reply:
This also might be a good use of the `sets` package since we really only care about the key (property type) and not the value in the map.

Reply:
@everettraven  I'm not a big fan of sets since they are just very slim wrapper around a no value map `map[T]struct{}{}`, but happy to switch to sets if people find it more clean. I'll do it in the next update.

> This might be a premature and low-value optimization, but it reduces the time complexity from O(n*m) to O(n)

Despite the nested loop this is considered O(n) since number of iterations in the inner loop is always 3 and does not depend on the input. Same as the following will be O(n):

```go
for i := range bundle.Properties {
	if bundle.Properties[i].Type == property.TypePackageRequired {
		return // ...
	}
	if bundle.Properties[i].Type == property.TypeGVKRequired {
		return // ...
	}
	if bundle.Properties[i].Type == property.TypeConstraint {
		return // ...
	}
}
```

Reply:
Fair enough on the time complexity. I guess I was approaching it from the perspective that the unsupported properties is an arbitrary value so while that is the case for this iteration it could shrink/grow as much or as little as we see fit. Using a map ensures that it is always an O(1) lookup.